code_tools/2 type_checkers/index.md

```diff
@@ -0,0 +1,570 @@
+
+<https://video.uva.nl/media/Platforms2021%20Typing%20C/0_4la7ngbg>
+
+
+    int sum(int items[], int n) {
+        int total = 0;
+        for (int i = 0; i < n; i++) {
+            total += items[i];
+        }
+        return total;
+    }
+
+
+C takes a different approach, put a concrete type in front of everything and check it when trying to compile. That way we'll know up front whether the code will even run. Because this:
+
+
+    float array[] = {3.0, 4.0, 5.0};
+    sum(array, 3);
+
+
+Will nicely throw a compile error. No chance that this code reaches the end user's desk. 
+
+But wait, floats can be summed right? Well, tough luck. You'll need to write a new function for floats.
+
+> For the curious, there are ways to escape C's type system through the use of casting and pointers. Most notably through the use of `void` pointers.
+
+</details>
+
+
+<details markdown="1"><summary  markdown="span">Dynamic vs Static</summary>
+
+All programming languages have some form of type system, but when and what they do with that system varies. First, let's talk about when. There are two main forms, **static** and **dynamic**, and they are not exclusive from one another. 
+
+#### Static
+
+![embed](https://api.eu.kaltura.com/p/120/sp/12000/embedIframeJs/uiconf_id/23449960/partner_id/120?iframeembed=true&playerId=kaltura_player&entry_id=0_lf6czbc1&flashvars[streamerType]=auto&amp;flashvars[localizationCode]=en_US&amp;flashvars[leadWithHTML5]=true&amp;flashvars[sideBarContainer.plugin]=true&amp;flashvars[sideBarContainer.position]=left&amp;flashvars[sideBarContainer.clickToClose]=true&amp;flashvars[chapters.plugin]=true&amp;flashvars[chapters.layout]=vertical&amp;flashvars[chapters.thumbnailRotator]=false&amp;flashvars[streamSelector.plugin]=true&amp;flashvars[EmbedPlayer.SpinnerTarget]=videoHolder&amp;flashvars[dualScreen.plugin]=true&amp;flashvars[hotspots.plugin]=1&amp;flashvars[Kaltura.addCrossoriginToIframe]=true&amp;&wid=0_yzi9tqnm)
+
+<https://video.uva.nl/media/Platforms2021%20Typing%20Static/0_lf6czbc1>
+
+Static in this context just means before execution, that could be when compiling the code or through running a seperate type checker. For instance, C makes use of static type checking to ensure that all types operate with one another upon compilation. That way, there is no (technically, little) chance for any type errors while running the program. On top of this compilers can make use of the type information upon compilation to better optimize the resulting program. By for instance reserving precisely enough memory, as the data and their types is know up front.
```

Comment:
"On top of this" -> voor mijn gevoel werkt "In addition, " beter. Bij "on top of this" krijg ik altijd een beetje het gevoel van een _negatieve_ opsomming ("hij ging óók nog op mijn tenen staan").